File: autoportfolio/api/schemas.py
"""Pydantic request/response models for the AutoPortfolio API."""
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class RecommendationRequest(BaseModel):
    portfolio_id: str
    current_holdings: dict[str, float] = Field(default_factory=dict)
    risk_appetite: Literal["conservative", "moderate", "aggressive"]
    capital: float = Field(gt=0)
    sector_constraints: list[str] = Field(default_factory=list)


class RecommendationResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    recommended_allocation: dict[str, float]
    expected_return: float
    expected_volatility: float
    confidence: float
    explanation: str
    model_version: str
    timestamp: str


class PortfolioStatusResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    portfolio_id: str
    model_version: str
    algorithm: str
    last_training_date: Optional[str]
    sharpe: float
    drift_score: float
    data_freshness_hours: float


class AllocationHistoryEntry(BaseModel):
    date: str
    allocation: dict[str, float]
    realized_return: Optional[float]


class PortfolioHistoryResponse(BaseModel):
    portfolio_id: str
    history: list[AllocationHistoryEntry]


class PipelineRunRequest(BaseModel):
    portfolio_id: str
    total_timesteps: int = 100_000
    n_trials: int = 20


class PipelineRunResponse(BaseModel):
    portfolio_id: str
    status: str
    message: str


class HealthResponse(BaseModel):
    status: str
    portfolios: list[str]
    mlflow: str


class PortfolioConfigResponse(BaseModel):
    name: str
    display_name: str
    tickers: list[str]
    risk_appetite: Literal["conservative", "moderate", "aggressive"]
    capital: float
    rebalance_frequency: str


class PortfolioListResponse(BaseModel):
    portfolios: list[PortfolioConfigResponse]


class PipelineRunStep(BaseModel):
    model_config = ConfigDict(extra="allow")

    step: str


class PipelineRunRecord(BaseModel):
    portfolio: str
    started_at: str
    steps: list[PipelineRunStep]
    status: str
    finished_at: Optional[str] = None


class PipelineRunsResponse(BaseModel):
    portfolio_id: str
    runs: list[PipelineRunRecord]


class PipelineLogEntry(BaseModel):
    seq: int
    timestamp: str
    level: str
    logger: str
    message: str


class PipelineLogsResponse(BaseModel):
    entries: list[PipelineLogEntry]

File: autoportfolio/frontend/lib/api/schemas.ts
import { z } from "zod";

const riskAppetiteSchema = z.enum(["conservative", "moderate", "aggressive"]);

export const recommendationResponseSchema = z.object({
  recommended_allocation: z.record(z.string(), z.number()),
  expected_return: z.number(),
  expected_volatility: z.number(),
  confidence: z.number(),
  explanation: z.string(),
  model_version: z.string(),
  timestamp: z.string(),
});

export const portfolioStatusResponseSchema = z.object({
  portfolio_id: z.string(),
  model_version: z.string(),
  algorithm: z.string(),
  last_training_date: z.string().nullable(),
  sharpe: z.number(),
  drift_score: z.number(),
  data_freshness_hours: z.number(),
});

export const allocationHistoryEntrySchema = z.object({
  date: z.string(),
  allocation: z.record(z.string(), z.number()),
  realized_return: z.number().nullable(),
});

export const portfolioHistoryResponseSchema = z.object({
  portfolio_id: z.string(),
  history: z.array(allocationHistoryEntrySchema),
});

export const pipelineRunResponseSchema = z.object({
  portfolio_id: z.string(),
  status: z.string(),
  message: z.string(),
});

export const healthResponseSchema = z.object({
  status: z.string(),
  portfolios: z.array(z.string()),
  mlflow: z.string(),
});

export const portfolioConfigResponseSchema = z.object({
  name: z.string(),
  display_name: z.string(),
  tickers: z.array(z.string()),
  risk_appetite: riskAppetiteSchema,
  capital: z.number(),
  rebalance_frequency: z.string(),
});

export const portfolioListResponseSchema = z.object({
  portfolios: z.array(portfolioConfigResponseSchema),
});

export const pipelineRunStepSchema = z
  .object({ step: z.string() })
  .catchall(z.unknown());

export const pipelineRunRecordSchema = z.object({
  portfolio: z.string(),
  started_at: z.string(),
  steps: z.array(pipelineRunStepSchema),
  status: z.string(),
  finished_at: z.string().nullable(),
});

export const pipelineRunsResponseSchema = z.object({
  portfolio_id: z.string(),
  runs: z.array(pipelineRunRecordSchema),
});

export const pipelineLogEntrySchema = z.object({
  seq: z.number(),
  timestamp: z.string(),
  level: z.string(),
  logger: z.string(),
  message: z.string(),
});

export const pipelineLogsResponseSchema = z.object({
  entries: z.array(pipelineLogEntrySchema),
});

export const algoProgressEntrySchema = z.object({
  phase: z.string(),
  trial: z.number(),
  n_trials: z.number(),
  steps_done: z.number(),
  steps_total: z.number(),
});

export const pipelineProgressResponseSchema = z.object({
  portfolio_id: z.string(),
  active: z.boolean(),
  fraction_done: z.number(),
  elapsed_seconds: z.number(),
  eta_seconds: z.number().nullable(),
  algorithms: z.record(z.string(), algoProgressEntrySchema),
});
